src/compose.ts

```diff
@@ -12,33 +14,41 @@ import type { Last } from './types/last'
  * matches the type of the previous function's argument.
  */
 export type Composition<
-  Functions extends Array<(...args: any[]) => any>,
+  Functions extends Array<Fn>,
   Length extends number = Functions['length']
   > =
   Length extends 1
-    ? Functions
-    : Functions extends [...infer Rest, infer Penultimate, infer Last]
-      ? [
-        ...Composition<
-          // Weirdly `Rest`, `Penultimate` and `Last`,
-          // which we extended from unary function type,
-          // are not considered as functions in this expression
-          // @ts-ignore
-          [...Rest, (arg: ReturnType<Last>) => ReturnType<Penultimate>]
-          >,
-        Last
-      ]
-      : any
+  ? Functions
+  : Functions extends [...infer Rest, infer Penultimate, infer Last]
+  ? [
+    ...Composition<
+      // Weirdly `Rest`, `Penultimate` and `Last`,
+      // which we extended from unary function type,
+      // are not considered as functions in this expression
+      Rest extends Array<Fn>
+      ? Last extends Fn
+      ? Penultimate extends Fn
+      ? [...Rest, (arg: ReturnType<Last>) => ReturnType<Penultimate>]
+      : never
+      : never
+      : never
+    >,
+    Last
+  ]
+  : Functions
 
 /**
  * Applies all of the provided `functions` one-by-one in right-to-left order
  * starting from the `argument`.
  */
-export default function compose<Functions extends Array<(arg: any) => any>> (
+export default function compose<Functions extends Array<Fn>>(
+  ...functions: Composition<Functions>
+): (arg: Parameters<Last<Functions>>[0]) => ReturnType<Functions[0]>
+export default function compose<Functions extends Array<Fn>>(
   ...functions: Composition<Functions>
-): (arg: Parameters<Last<Functions>>[0]) => ReturnType<Functions[0]> {
-  return (arg) => {
-    const length = (functions as any[]).length
+) {
+  return (arg: Parameters<Last<Functions>>[0]) => {
```

Comment:
This type signature is excessive as we've explicitly annotated the `arg` on line 47.
Actually, this style of annotations is seen across the whole library – adding types before actual runtime code to make the declarations clear from the start.

Reply:
There is a problem with assignability. `Parameters<Last<Functions>>[0]` is treated as `unknown`